garak/probes/topic.py

```diff
@@ -0,0 +1,206 @@
+            # this is one attempt
+            # mint attempt
+            # do generation
+            responses = generate(prompt)
+            scores = [detector_score(response) for response in responses]
+            # send to detector...
+            yield prompt, scores
+
+    def probe(self, generator):
+        # open_terms = []
+        # closed_terms = []
+
+        self.generator = generator
+        detector = garak._plugins.load_plugin(self.detector_plugin)
+
+        all_completed_attempts: Iterable[garak.attempt.Attempt] = []
+
+        while len(self.synsets_to_explore):
+            logging.debug(f"{self.__class__.__name__} Queue: {self.synsets_to_explore}")
+            current_synset = self.synsets_to_explore.pop()
+            self.synsets_ids_explored.add(current_synset.id)
+            # go through all words in synset
+            # if any stick, add children to queue
+            # results = []
+
+            # build list of attempts
+            attempts_todo: Iterable[garak.attempt.Attempt] = []
+
+            logging.debug(
+                f"{self.__class__.__name__} {current_synset}, {current_synset.words()}"
+            )
+            for word in current_synset.words():
+                for surface_form in word.forms():
+                    for prompt in self._gen_prompts(surface_form):
+                        a = self._mint_attempt(prompt)
+                        a.notes["surface_form"] = surface_form
+                        attempts_todo.append(a)
+
+            # buff hook
+            if len(_config.buffmanager.buffs) > 0:
+                attempts_todo = self._buff_hook(attempts_todo)
+
+            attempts_completed = self._execute_all(attempts_todo)
+            all_completed_attempts += attempts_completed
+
+            # now we call the detector 🙃
+            synset_results = []
+            for a in attempts_completed:
+                synset_results += detector.detect(a)
+
+            synset_results = [
+                1.0 if s > self.PER_GENERATION_THRESHOLD else 0 for s in synset_results
+            ]
+
+            mean_score = sum(synset_results) / len(synset_results)
+            logging.debug(f"{self.__class__.__name__}  synset score {mean_score}")
+
+            if mean_score > self.PER_SYNSET_THRESHOLD:
+                children = current_synset.hyponyms()
+                logging.debug(
+                    f"{self.__class__.__name__}  adding hyponyms - children"
+                    + repr(children)
+                )
+                for child in children:
+                    if child.id not in self.synsets_ids_explored:
+                        logging.debug(f"{self.__class__.__name__}   {child}")
+                        self.synsets_to_explore.append(child)
+                    else:
+                        logging.debug(f"{self.__class__.__name__}   skipping {child}")
+            else:
+                logging.debug(f"{self.__class__.__name__} closing synset node")
```

Comment:
How does running a configuration provided detector impact later processing by the `primary_detector`, should these values be required to be the same detector?

Are we evaluating the Attempt results twice possibly with the same detector?

Reply:
Yeah, double eval. I considered adding a passthru detector, to avoid this. Not 100% sure how this will be used, but I'm leaning toward the final detection not being so useful.

Ideally I'd like the detector used during running to be specified in `primary_detector`. The harness will invoke this again when the probe completes; it might be possible for the probe to switch out its detector value after probing is complete, to something like `always.Passthru`. Doing this introduces an assumption about how harness works, and that seems an worth codifying/making explicit in a test. Not sure how that test would like. Will try out the passthru substitution method and think on it.

Reply:
addressed using passthru and some switching out at probe completion